### ISSUE TITLE
Error: SortDetails.cpp, 885: Unable to create BED structure.

### ISSUE THREAD

**Issue by alexlenail** [bedops/bedops#265]
https://github.com/bedops/bedops/blob/abf5c517d921cf1ec188119293b021491e891e66/applications/bed/sort-bed/src/SortDetails.cpp#L883-L885

What should I look for in my bed file, which might be causing this error?

**Comment by alexpreynolds** [comment]
Interesting error. 

At a first glance, if a (signed) `int` is negative, then that would suggest overflow — more than `INT_MAX` or 2147483647 distinct BED elements (lines)? (Are you working with a large file of SNPs?)

https://github.com/bedops/bedops/blob/abf5c517d921cf1ec188119293b021491e891e66/applications/bed/sort-bed/src/SortDetails.cpp#L356

Can you provide the output from `wc -l in.bed`, just to provide a glance at how many lines are in your data?

It might also be worth looking into `--max-mem` and `--tmpdir` options to sort via disk, using smaller chunks of memory, e.g. `2G` or the like.

**Comment by alexpreynolds** [comment]
Patching line 883 might make sense, but I'd still be curious to know more about the file you're trying to sort.

**Comment by alexlenail** [comment]
Thanks for getting back, @alexpreynolds 

My largest bed file is 200GB. One of the ~120GB files has 2342747652 lines, which is greater than INT_MAX. 

What's the easiest way to patch that? I'm running this on a cluster on 1TB RAM nodes, so memory isn't as much of an issue..?

**Comment by alexpreynolds** [comment]
It may be enough to change line 883 from:

```
if (static_cast<int>(chromEntryCount) < 0) 
```

to:

```
if (static_cast<Bed::SignedCoordType>(chromEntryCount) < 0) 
```

and recompiling. 

This patch would use a 64-bit signed integer type under the hood (`int64_t`), which I expect should avoid the overflow condition observed with the 32-bit signed integer (`int`). The maximum `int64_t` is 9223372036854775807. 

I would need to test this to be sure, but that's where I will probably start.

In the meantime, I would recommend sorting on disk in chunks, via an option pair like `--max-mem 2G --tmpdir ${PWD}` or similar. I haven't tested this out, either, but sorting in smaller chunks with fewer lines — while it will take longer — may not result in the overflow condition. Ideally the patch above would avoid this issue and permit sorting in-memory, as you have that memory available to you, presumably.

**Comment by alexlenail** [comment]
I just deleted 883-887

https://github.com/bedops/bedops/blob/abf5c517d921cf1ec188119293b021491e891e66/applications/bed/sort-bed/src/SortDetails.cpp#L883-L887

built and ran, and it seems to work now (exits successfully -- and files look sorted)

**Comment by alexpreynolds** [comment]
(I'm reopening this, so that I can track this for the next revision.)

**Comment by alexpreynolds** [comment]
Patched in commit d0461223
